### ISSUE TITLE
user experience after version rollback

### ISSUE THREAD

**Issue by mikesteele81** [lxc/lxd#2299]
This is a feature(s) request.

I've been using the ubuntu-lxc/lxd-stable PPA. This morning I allowed an upgrade from 2.0.4 to 2.1.0, and then decided I'd rather stay on an LTS version after all.

After rolling back my version to 2.0.3, none of `lxc`'s subcommands would work. The command prompt would freeze. I attempted to restart the `lxd` service with `sudo service lxd restart`, and once again the command just sat there until manually killed.

By checking syslog I was able to see what the problem was:

    Aug 19 09:15:20 ubuntu01 systemd[1]: Starting LXD - unix socket.
    Aug 19 09:15:20 ubuntu01 systemd[1]: Listening on LXD - unix socket.
    Aug 19 09:15:20 ubuntu01 systemd[1]: Starting LXD - main daemon...
    ...
    Aug 19 09:15:20 ubuntu01 lxd[5489]: t=2016-08-19T09:15:20-0700 lvl=warn msg="CGroup memory swap accounting is disabled, swap limits will be ignored."
    Aug 19 09:15:20 ubuntu01 lxd[5489]: error: Bad database version: 33
    Aug 19 09:15:20 ubuntu01 systemd[1]: lxd.service: Main process exited, code=exited, status=1/FAILURE

Ignoring what should be done about database version mismatches, I think the following features would be nice to have:

* Internal failures in starting the `lxd` service should cause a manual systemd service start to fail as well. Currently it just continues to wait for the service to start.
* `lxc` sub-commands should return a failure message when the `lxd` service is in a failed state due to a database version mismatch. Currently they just sit there.
* When the `lxd` service is stopped, `lxc` sub-commands should report that they cannot run unless the service is started. Currently the reported error says, `Permission denied, are you in the lxd group?`

Regarding the database, I wish that it had not underwent a schema update implicitly as part of upgrading the lxd software. It would be nice if new versions of lxd were able to function in a compatibility mode with the old database until it was upgraded manually under direction of the user.

**Comment by stgraber** [comment]
# systemd
So the systemd part is a distro issue and not tracked upstream, so we can ignore this here. I'm not sure there's much we can do in systemd anyway given that the LXD daemon does start, then fail very early on, exiting with an error, which systemd does notice and log (as seen above).

The problem appears to be that it doesn't then cancel the post-start actions it started, leading to that waitstart process still being around.

# client interacting with a LXD which isn't running yet
LXD does actually know how to differentiate between a socket that's not bound and one that can't be talked to because of permission issues.

    stgraber@castiana:~$ lxc list
    LXD socket not found; is LXD installed and running?

    stgraber@castiana:~$ sudo -u nobody -- lxc list
    Permission denied, are you in the lxd group?

The problem, here again is systemd. Since LXD is socket activated in your case, the socket is always present and bound. Systemd then prevents you from connecting to it if the daemon doesn't respond, which is then indistinguishable from a permission issue.

There may however be an issue for when the socket exists and isn't bound (which is something that LXD itself never does). I'll look into that one.

# Telling the user about schema update failures

We absolutely don't want to ever have the daemon running with a database it doesn't understand.
So the best we can do is log an error, which we do already.

Maintaining schema backward compatibility isn't something we're interested in, it's way too costly to do. Same goes with keeping the code compatible with older schemas. We don't do schema changes for fun, we do them because we need them, if there's a way to add a feature without changing the schema, that's what we'll do.

# Allowing for downgrades

So while we won't ever be supporting it, one thing we can definitely do to make downgrades slightly more likely to work is to make a copy of the database before doing any schema upgrade.
If something goes wrong or if you want to downgrade, you can then grab the "lxd.db.bak" file and copy it over "lxd.db" before starting the old version.

Note that this only works so long as the only thing we changed during version upgrade is DB schema. We've been known to also move various filesystem bits around in the past, which you'd then need to manually fix (shouldn't apply to 2.0.x to 2.1 though).

**Comment by stgraber** [comment]
Ok, so for the permission denied error, there's nothing we can do as it's the kernel telling us permission denied when it's in fact "unconnected socket":
```
stgraber@castiana:~/data/code/lxc/lxd (stgraber/master)$ nc -U /var/lib/lxd/unix.socket 
nc: unix connect failed: Permission denied
```

I'm looking at doing the DB backup bit now.

**Comment by mikesteele81** [comment]
Thank you for considering all of this. I'll feel better knowing an automatic backup of the database gets made when a schema update happens. If the `lxc` command-line interface cannot distinguish between a closed socket and a permissions error, would it make sense to give a generic error message such as, "Unable to communicate with LXD" rather than a specific but possibly misleading error message?

I decided to go ahead and use the 2.1 release. Future updates will likely contain improvements which I'll be eager to use.
